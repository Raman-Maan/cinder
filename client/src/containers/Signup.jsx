import React, { Component } from 'react';
import { Container, Form, FormGroup, Label, Input, Button } from 'reactstrap';
import { Redirect, Link } from 'react-router-dom';

import './styles/Signup.css';
import logo from '../assets/logo.svg';

class Signup extends Component {
  constructor(props) {
    super(props);

    this.setInput = this.setInput.bind(this);
    this.fetchGenderList = this.fetchGenderList.bind(this);
    this.userSignup = this.userSignup.bind(this);

    this.state = {
      genderList: [],
      email: '',
      emailConfirm: '',
      emailCheck: false,
      password: '',
      passwordConfirm: '',
      userName: '',
      birthday: '',
      gender: '',
      signnedUp: false,
      signupBtnText: 'Sign up'
    };
  }

  componentDidMount() {
    this.fetchGenderList();
  }

  fetchGenderList() {
    this.setState({
      genderList: [
        {
          id: 0,
          type: 'Male'
        },
        {
          id: 1,
          type: 'Female'
        },
        {
          id: 2,
          type: 'Other'
        }
      ]
    });
  }

  setInput(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  userSignup(e) {
    e.preventDefault();

    this.setState({ signnedUp: true });
  }

  render() {
    if (this.state.signnedUp) {
      return <Redirect to="/login" />;
    }

    return (
      <div className="Signup">
        <Container>
          <div className="logo">
            <h1 className="display-3">cinder</h1>
            <img src={logo} className="App-logo" alt="logo" />
          </div>
          <hr />

          <div className="inputPart">
            <Form className="signup-form" onSubmit={e => this.userSignup(e)}>
              <div className="row">
                <div className="col-sm">
                  <FormGroup>
                    <Label for="email">Email</Label>
                    <Input
                      required
                      id="email"
                      type="email"
                      name="email"
                      onChange={this.setInput}
                    />
                  </FormGroup>
                </div>
                <div className="col-sm">
                  <FormGroup>
                    <Label for="emailConfirm">Please re-enter email</Label>
                    <Input
                      required
                      id="emailConfirm"
                      type="email"
                      name="emailConfirm"
                      onChange={this.setInput}
                    />
                  </FormGroup>
                </div>
              </div>

              <div className="row">
                <div className="col-sm">
                  <FormGroup>
                    <Label for="password">Password</Label>
                    <Input
                      required
                      id="password"
                      type="password"
                      name="password"
                      onChange={this.setInput}
                    />
                  </FormGroup>
                </div>
                <div className="col-sm">
                  <FormGroup>
                    <Label for="passwordConfirm">
                      Please re-enter password
                    </Label>
                    <Input
                      required
                      id="passwordConfirm"
                      type="password"
                      name="passwordConfirm"
                      onChange={this.setInput}
                    />
                  </FormGroup>
                </div>
              </div>

              <div className="row">
                <div className="col-sm">
                  <FormGroup>
                    <Label for="gender">Gender</Label>
                    <Input
                      type="select"
                      name="gender"
                      id="gender"
                      required
                      onChange={this.setInput}
                    >
                      <option value="" />
                      {this.state.genderList.map(opt => {
                        return (
                          <option key={opt.id} value={opt.id}>
                            {opt.type}
                          </option>
                        );
                      })}
                    </Input>
                  </FormGroup>
                </div>
                <div className="col-sm">
                  <FormGroup>
                    <Label for="birthday">Birthday</Label>
                    <Input
                      required
                      type="date"
                      name="birthday"
                      id="birthday"
                      onChange={this.setInput}
                    />
                  </FormGroup>
                </div>
              </div>

              <FormGroup>
                <Label for="userName">Name</Label>
                <Input
                  required
                  id="userName"
                  name="userName"
                  placeholder="Just first name is enough"
                  onChange={this.setInput}
                />
              </FormGroup>

              <FormGroup>
                <Button id="submitBtn" outline block color="dark" type="submit">
                  {this.state.signupBtnText}
                </Button>

                <Button
                  id="login"
                  outline
                  block
                  color="dark"
                  tag={Link}
                  to="/login"
                >
                  Log in
                </Button>
              </FormGroup>
            </Form>
          </div>
        </Container>
      </div>
    );
  }
}

export default Signup;
