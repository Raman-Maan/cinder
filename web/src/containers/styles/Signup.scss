.Signup {
  display: flex;
  height: 100%;
  width: 100%;

  justify-content: center;
  align-items: center;

  .container {
    .logo {
      display: flex;
      justify-content: center;
      align-content: center;

      img {
        animation: logo-spin infinite 20s linear;
        height: 80px;
      }
    }
  }
}

@keyframes logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
