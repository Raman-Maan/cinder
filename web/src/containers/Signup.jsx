import React, { Component } from 'react';
import {
  Alert,
  Container,
  Form,
  FormGroup,
  Label,
  Input,
  Button,
  Row,
  Col
} from 'reactstrap';
import { signupUser } from '../actions';
import { connect } from 'react-redux';
import { Redirect, Link } from 'react-router-dom';

import './styles/Signup.css';
import logo from '../assets/logo.svg';

export class Signup extends Component {
  constructor(props) {
    super(props);

    this.setInput = this.setInput.bind(this);
    this.fetchGenderList = this.fetchGenderList.bind(this);
    this.userSignup = this.userSignup.bind(this);

    this.state = {
      errored: false,
      message: '',
      genderList: [],
      email: '',
      emailConfirm: '',
      password: '',
      passwordConfirm: '',
      userName: '',
      birthday: '',
      gender: ''
    };
  }

  componentDidMount() {
    this.fetchGenderList();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      errored: nextProps.errored,
      message: nextProps.message
    });
  }

  fetchGenderList() {
    this.setState({
      genderList: [
        {
          id: 1,
          type: 'Male'
        },
        {
          id: 2,
          type: 'Female'
        },
        {
          id: 3,
          type: 'Other'
        }
      ]
    });
  }

  setInput(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  userSignup(e) {
    e.preventDefault();
    if (
      this.state.email === this.state.emailConfirm &&
      this.state.password === this.state.passwordConfirm
    ) {
      this.props.signupUser({
        email: this.state.email,
        password: this.state.password,
        userName: this.state.userName,
        birthday: this.state.birthday,
        genderID: this.state.gender
      });
    } else {
      this.setState({
        errored: true,
        message: 'Your email or password are not matching'
      });
    }
  }

  render() {
    if (this.props.isAuthenticated) {
      const { from } = this.props.location.state || { from: { pathname: '/' } };
      return <Redirect exact to={from} />;
    }

    return (
      <div className="Signup">
        <Container>
          <div className="logo">
            <h1 className="display-3">cinder</h1>
            <img src={logo} className="App-logo" alt="logo" />
          </div>
          <hr />
          {this.state.errored && (
            <Alert color="danger">{this.state.message}</Alert>
          )}
          <Form className="signup-form" onSubmit={e => this.userSignup(e)}>
            <Row>
              <Col sm={6}>
                <FormGroup>
                  <Label for="email">Email</Label>
                  <Input
                    required
                    id="email"
                    type="email"
                    name="email"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label for="emailConfirm">Please re-enter email</Label>
                  <Input
                    required
                    id="emailConfirm"
                    type="email"
                    name="emailConfirm"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
            </Row>

            <Row>
              <Col sm={6}>
                <FormGroup>
                  <Label for="password">Password</Label>
                  <Input
                    required
                    id="password"
                    type="password"
                    name="password"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label for="passwordConfirm">Please re-enter password</Label>
                  <Input
                    required
                    id="passwordConfirm"
                    type="password"
                    name="passwordConfirm"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
            </Row>

            <Row>
              <Col sm={6}>
                <FormGroup>
                  <Label for="gender">Gender</Label>
                  <Input
                    type="select"
                    name="gender"
                    id="gender"
                    required
                    onChange={this.setInput}
                  >
                    <option value="" />
                    {this.state.genderList.map(opt => {
                      return (
                        <option key={opt.id} value={opt.id}>
                          {opt.type}
                        </option>
                      );
                    })}
                  </Input>
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label for="birthday">Birthday</Label>
                  <Input
                    required
                    type="date"
                    name="birthday"
                    id="birthday"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
            </Row>

            <Row>
              <Col sm={12}>
                <FormGroup>
                  <Label for="userName">Name</Label>
                  <Input
                    required
                    id="userName"
                    name="userName"
                    placeholder="Just first name is enough"
                    onChange={this.setInput}
                  />
                </FormGroup>
              </Col>
            </Row>

            <Row>
              <Col sm={6}>
                <Button
                  id="submitBtn"
                  outline
                  color="dark"
                  type="submit"
                  size="md"
                  className="float-right"
                >
                  Sign Up
                </Button>
              </Col>

              <Col sm={6}>
                <Button
                  id="login"
                  color="dark"
                  to="/login"
                  tag={Link}
                  outline
                  size="md"
                >
                  Go back to Log In
                </Button>
              </Col>
            </Row>
          </Form>
        </Container>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  message: state.auth.message,
  isAuthenticated: state.auth.isAuthenticated,
  errored: state.auth.errored
});

const mapDispatchToProps = dispatch => ({
  signupUser: creds => dispatch(signupUser(creds))
});

export default connect(mapStateToProps, mapDispatchToProps)(Signup);
