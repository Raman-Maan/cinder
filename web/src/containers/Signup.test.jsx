import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Signup } from './Signup';

describe('<Signup />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<Signup />);
  });

  it('renders without crashing', () => {
    // simple smoke test
    shallow(
      <MemoryRouter>
        <Signup />
      </MemoryRouter>
    );
  });

  it('renders correctly', () => {
    const component = renderer
      .create(
        <MemoryRouter>
          <Signup />
        </MemoryRouter>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  describe('Form value', () => {
    let testValues;

    beforeEach(() => {
      wrapper = shallow(<Signup />);
      testValues = {
        email: 'foo@email.com',
        emailConfirm: 'foo@email.com',
        password: 'something',
        passwordConfirm: 'something',
        userName: 'foo',
        birthday: '1990-01-02',
        gender: '0',
        handleSubmit: jest.fn()
      };
    });

    it('changes the form value accordingly to input', () => {
      wrapper.find('#email').simulate('change', {
        target: { name: 'email', value: testValues.email }
      });

      expect(wrapper.state().email).toBe(testValues.email);
    });
  });

  describe('Submit sign up', () => {
    const mockSignup = jest.fn();
    let testValues;

    beforeEach(() => {
      wrapper.instance().userSignup = mockSignup;
      wrapper.update();
      mockSignup.mockClear();

      testValues = {
        email: 'foo@email.com',
        emailConfirm: 'foo@email.com',
        password: 'something',
        passwordConfirm: 'something',
        userName: 'foo',
        birthday: '1990-01-02',
        gender: 'Male',
        handleSubmit: jest.fn()
      };
    });

    it('calls the mock function when form is submitted', () => {
      wrapper.find('.signup-form').simulate('submit', { preventDefault() {} });
      expect(mockSignup.mock.calls.length).toBe(1);
    });

    it('does not submit with empty fields', () => {
      wrapper.find('#submitBtn').simulate('click');
      expect(wrapper.find('.Signup')).toHaveLength(1);
      expect(mockSignup).toHaveBeenCalledTimes(0);
    });

    // TODO
    // it('submits with complete fields', () => {
    // });
  });
});
