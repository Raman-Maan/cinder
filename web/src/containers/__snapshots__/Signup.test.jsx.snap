// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders correctly 1`] = `
<div
  className="Signup"
>
  <div
    className="container"
  >
    <div
      className="logo"
    >
      <h1
        className="display-3"
      >
        cinder
      </h1>
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </div>
    <hr />
    <form
      className="signup-form"
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="email"
            >
              Email
            </label>
            <input
              className="form-control"
              id="email"
              name="email"
              onChange={[Function]}
              required={true}
              type="email"
            />
          </div>
        </div>
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="emailConfirm"
            >
              Please re-enter email
            </label>
            <input
              className="form-control"
              id="emailConfirm"
              name="emailConfirm"
              onChange={[Function]}
              required={true}
              type="email"
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="password"
            >
              Password
            </label>
            <input
              className="form-control"
              id="password"
              name="password"
              onChange={[Function]}
              required={true}
              type="password"
            />
          </div>
        </div>
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="passwordConfirm"
            >
              Please re-enter password
            </label>
            <input
              className="form-control"
              id="passwordConfirm"
              name="passwordConfirm"
              onChange={[Function]}
              required={true}
              type="password"
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="gender"
            >
              Gender
            </label>
            <select
              className="form-control"
              id="gender"
              name="gender"
              onChange={[Function]}
              required={true}
            >
              <option
                value=""
              />
              <option
                value={1}
              >
                Male
              </option>
              <option
                value={2}
              >
                Female
              </option>
              <option
                value={3}
              >
                Other
              </option>
            </select>
          </div>
        </div>
        <div
          className="col col-sm-6"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="birthday"
            >
              Birthday
            </label>
            <input
              className="form-control"
              id="birthday"
              name="birthday"
              onChange={[Function]}
              required={true}
              type="date"
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col col-sm-12"
        >
          <div
            className="form-group"
          >
            <label
              className=" form-control-label"
              htmlFor="userName"
            >
              Name
            </label>
            <input
              className="form-control"
              id="userName"
              name="userName"
              onChange={[Function]}
              placeholder="Just first name is enough"
              required={true}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col col-sm-6"
        >
          <button
            className="float-right btn btn-outline-dark btn-md"
            id="submitBtn"
            onClick={[Function]}
            type="submit"
          >
            Sign Up
          </button>
        </div>
        <div
          className="col col-sm-6"
        >
          <a
            className="btn btn-outline-dark btn-md"
            href="/login"
            id="login"
            onClick={[Function]}
            type={undefined}
          >
            Go back to Log In
          </a>
        </div>
      </div>
    </form>
  </div>
</div>
`;
